ObjectSeeker Defense Implementation

f4str opened the pull request:
# Description

Implement the ObjectSeeker certifiably robust defense for object detectors. This is an estimator located in `art.estimators.certification.object_seeker` as the `PyTorchObjectSeeker` class. It is constructed in a similar way as the `PyTorchObjectDetector` or `PyTorchYolo`. Currently, there is only a PyTorch implementation, but a generic and TensorFlow implementation is possible in the future.

Additionally created utility functions for `intersection_over_union`, `intersection_over_area`, and `non_maximum_supression` under `art.utils`.

Note that unit tests are in `tests/estimators/object_detection` rather than `tests/estimators/certification` since a YOLO model is required.

Relevant links:
* Paper: <https://arxiv.org/abs/2202.01811>
* Code: <https://github.com/inspire-group/ObjectSeeker>

Fixes #2193 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit tests for ObjectSeeker YOLO
- [x] Unit tests for ObjectSeeker Faster R-CNN

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

beat-buesser commented on tests/estimators/object_detection/test_object_seeker.py:
```diff
@@ -0,0 +1,186 @@
+
+
+@pytest.fixture()
+def get_pytorch_yolo(get_default_cifar10_subset):
```
Is this a fixture that we also could use in other tests and move to contest.py?

beat-buesser commented on tests/estimators/object_detection/test_object_seeker.py:
```diff
@@ -0,0 +1,186 @@
+        # Compute loss after training
+        loss2 = object_seeker.compute_loss(x=x_test, y=y_test)
+
+        assert loss1 != loss2
```
Should `loss2` be smaller than `loss1`?

f4str replied:
I borrowed the same unit test from `test_pytorch_yolo.py`. Logically it should be less than, but I believe it was not equal in case of stochasticity that causes the loss to increase. I can look into this and see if it can be changed to less than for both tests.

beat-buesser reviewed (approved):
Hi @f4str Thank you very much for your well prepared pull request implementing ObjectSeeker in ART! I have only a few questions and suggestions, but otherwise the code looks good.

f4str commented on art/estimators/certification/object_seeker/object_seeker.py:
```diff
@@ -0,0 +1,379 @@
+class ObjectSeekerMixin(abc.ABC):
+    """
+    Implementation of the ObjectSeeker certifiable robust defense applied to object detection models.
+    The original implementation is https://github.com/inspire-group/ObjectSeeker
```
The original code has an MIT license, so we can probably include their license below the ART one.

f4str commented:
Hi @beat-buesser thank you for the review. I've addressed the changes by adding the license from the original ObjectSeeker repo and extracting the fixtures into a `conftest.py`. I've extracted both the YOLO and Faster R-CNN fixtures and created a new unit test for Faster R-CNN. Now ObjectSeeker will test both object detector models.

As for the training loss not-equal vs less-than, it needs to remain a not-equal condition since due to stochasticity, the loss does not always decrease, especially with the "incorrect" dataset. The same applies to the PyTorchYolo and PyTorchFasterRCNN models.

beat-buesser commented:
Hi @f4str Thank you very much for the updates as well as checking and documenting! We'll merge this PR as soon as possible.
